app/scripts/controllers/edit/buildConfig.js

```diff
@@ -348,6 +375,77 @@ angular.module('openshiftConsole')
       return triggers;
     };
 
+    var loadBuildConfigSecrets = function(buildConfig) {
+      $scope.secrets.picked = {
+        gitSecret: normalizeSourceSecretFormat($scope.buildConfig.spec.source.sourceSecret),
+        pullSecret: normalizeSourceSecretFormat(buildStrategy($scope.buildConfig).pullSecret),
+        pushSecret: normalizeSourceSecretFormat($scope.buildConfig.spec.output.pushSecret),
+      }
+
+      if ($scope.strategyType === "Source" || $scope.strategyType === "Docker") {
+        $scope.secrets.picked.sourceSecrets = normalizeSourceSecretFormat($scope.buildConfig.spec.source.secrets, "secret", "destinationDir");
+      } else if ($scope.strategyType === "Custom") {
+        $scope.secrets.picked.sourceSecrets = normalizeSourceSecretFormat(buildStrategy($scope.buildConfig).secrets, "secretSource", "mountPath");
+      }
+    }
+
+    // normalizeSourceSecretFormat will transform secrets of all types into a unified format eg:
+    //  pullSecret original format:   {name: 'pullSecretName'}
+    //  pullSecret normalized format: [{name: 'pullSecretName'}]
+    //
+    //  secretSpec original format:   [{secretSource: {name: "secretSpecName"}, mountPath: "path"}]
+    //  secretSpec normalized format: [{name: "secretSpecName", secondaryInput: "path"}]
+    //
+    var normalizeSourceSecretFormat = function(secrets, secretFieldName, sourceTo) {
```

Comment:
Is normalize/denormalize really necessary? Once you start transforming things to and from from original format, it's easy to lose data. I'd prefer to work with the API objects as we get them from the server.